### PR TITLE
feat: instrumentation for racecar

### PR DESCRIPTION
**Pull request by chrisholmes**
This PR implements https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/15

Racecar is built on top of rdkafka, but it doesn't use the API that is implemented by the rdkafka instrumentation so an racecar specific adapter is required. 

Racecar's [main loop](https://github.com/zendesk/racecar/blob/73f424f36324f4161f820f6558097a704cbb3316/lib/racecar/runner.rb#L66-L80) repeatedly calls a rdkafka consumer's `#poll` method and then sends the returned messages to its own consumer classes (aka processors).

This instrumentation works by patching the Runner class and then its processor (Racecar::Consumer objects) at initialization.

### CODE REVIEW

**Comment by plantfansam** [comment]
Thanks @chrisholmes! Are you set up with a tracing UI locally? If so, would it be possible to include a screenshot of a `racecar` trace? This isn't part of our PR template (yet!), but it might be nice to see what this looks like!

**Comment by chrisholmes** [comment]
@plantfansam yep, makes sense. Here's the view in jaeger:

image

**Review by arielvalentin — COMMENTED** [review]
Thanks again for this submission! 

I am curious about the choice to mixin the instrumentation separately of the processor object vs mixing into the built in instrumenter[^1] or decorating `process`, and `process_batch`.

Would you mind sharing your thought process behind the overall design? 

[^1]: https://github.com/zendesk/racecar/blob/master/lib/racecar/instrumenter.rb

**Comment by chrisholmes** [comment]
> Thanks again for this submission!
> 
> I am curious about the choice to mixin the instrumentation separately of the processor object vs mixing into the built in instrumenter[1](#user-content-fn-1-5f7e614a36a1a95ce99b518b507ebc18) or decorating `process`, and `process_batch`.
> 
> Would you mind sharing your thought process behind the overall design?
> 
> ## Footnotes
> 1. https://github.com/zendesk/racecar/blob/master/lib/racecar/instrumenter.rb [↩](#user-content-fnref-1-5f7e614a36a1a95ce99b518b507ebc18)

I've tried a few different ways, but this form (though it is quite ugly) is the only format I can get the tests to pass so its spans are consistent to the other kafka instrumentations. 

This is my understanding of the impact of the methods you've mentioned:

When patching `process` and `process_batch` on `Racecar::Runner` the tests will pass except that exceptions are not propagated up to the Otel layer when racecar has its `pause` functionality enabled. Frustratingly, [`with_pause`](https://github.com/zendesk/racecar/blob/7a4bbcfbd173d4ce0323e3aae7acaca58ab4ed29/lib/racecar/runner.rb#L191) doesn't propagate the exception it rescues.

As far as I can tell, it is not fully possible to adapt Racecar's instrumentation interface as the `process_batch` payload doesn't have access to the batch of messages or their headers so it is not possible to add links to the traces. The `process_message` payload does have access to the message headers so we could use it there.

What do you think about this? Is it worth omitting batch message links or errors for a simpler looking instrumentation?

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (chrisholmes) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb]: sure, let me know if I'd done what you'd expect.

**Comment by ericmustin** [comment]
assigned myself to this so i can remember to review while ariel is out, apols on the delay here chris.

File: instrumentation/all/Gemfile
Comment (fbogsany) [inline instrumentation/all/Gemfile:11]: Why is this added here? It isn't necessary for the other instrumentation gems.

File: instrumentation/racecar/example/Gemfile
Comment (fbogsany) [inline instrumentation/racecar/example/Gemfile:14]: Is this line necessary for the example?

File: instrumentation/racecar/example/tracing.rb
Comment (fbogsany) [inline instrumentation/racecar/example/tracing.rb:4]: Do we need this line?

File: instrumentation/all/Gemfile
Comment (chrisholmes) [reply instrumentation/all/Gemfile]: actually, I'm wrong. This line is required so that the we can add the new instrumentation to all before it has been released. If we remove that line there are build errors as the all gem doesn't have a version of this gem to reference.

File: instrumentation/racecar/example/tracing.rb
Comment (chrisholmes) [reply instrumentation/racecar/example/tracing.rb]: you're right, we don't need it

File: instrumentation/racecar/example/Gemfile
Comment (chrisholmes) [reply instrumentation/racecar/example/Gemfile]: no

File: instrumentation/racecar/Appraisals
Comment (ericmustin) [inline instrumentation/racecar/Appraisals:15]: As noted, this needs to be updated to run appraisals for racecar

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (ericmustin) [inline instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb:17]: Nit but would you mind abstracting this gem version check into a private method `gem_version`, which `compatible` then calls ? This follows the style we've tried to implement in this pr https://github.com/open-telemetry/opentelemetry-ruby-contrib/commit/3b03ff7ea66b69c85ba205a369b85c2c33b712fe

see example https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/26a9b3b7cc6e37f71b29087ee06639aad6cd459a/instrumentation/pg/lib/opentelemetry/instrumentation/pg/instrumentation.rb#L50-L52

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (ericmustin) [inline instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb:21]: I'm a bit confused here. So is the idea that the racecar instrumentation depends on the rdkafka instrumentation?

1. Is it possible to check if the instrumentation is already installed before attempting to re-install it?
2. What's the desired behavior if the rdkafka instrumentation has been intentionally disabled by the user via env var? Are we essentially overwriting that choice.
2a. Hypothetical since there's no config options at the moment, but won't we automatically overwriting any configuration the user has supplied for their `rdkafka` instrumentation should it be aded in the future?

I guess i have to think a bit more closely about what the "right" behavior should be, but imo we should be respecting if this instrumentation has been disabled and, imo, its not clear to me why we need to take rdkafka as a dependancy.

File: instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec
Comment (ericmustin) [inline instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec:30]: why `~> 0.2.0`? This is lower than the min version of our rdkafka instrumentation (0.10.0) https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/26a9b3b7cc6e37f71b29087ee06639aad6cd459a/instrumentation/rdkafka/lib/opentelemetry/instrumentation/rdkafka/instrumentation.rb#L12

Should we be making the rdkafka instrumentation more permissive? As mentioned before, currently the coupling between these two instrumentations feels unclear and maybe a little bit unnecessary

**Review by ericmustin — CHANGES_REQUESTED** [review]
Overall this looks good.

There are a few TODOs you have that still need to be completed,

The only real blocker to me is it's not clear to me why we need to take `rdkafka` as a hard dependency and install the instrumentation, I'd like to understand this better and whether it would be possible to decouple them. If not, I'd like to make sure we respect any env var's that disable the rdkafka instrumentation as well as make sure there's no edge cases in version requirements. 

Nice work as always, apols on the delay on review

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (chrisholmes) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb:21]: Yes, my intention here was to rely on the Rdkafka implementation of producer spans as adding a Racecar implementation would duplicate them like this:

```
racecar send span:
  rdkafka send span:
    consumer span
```

There is a similar dependency between the Rack instrumentation with Rails and Sinatra so I did assume it would be alright to follow that pattern, but it is being used slightly different here and less extensively.

> 1. Is it possible to check if the instrumentation is already installed before attempting to re-install it?

Yes, there is an interface we could use.

> 2. What's the desired behavior if the rdkafka instrumentation has been intentionally disabled by the user via env var? Are we essentially overwriting that choice.
>    2a. Hypothetical since there's no config options at the moment, but won't we automatically overwriting any configuration the user has supplied for their `rdkafka` instrumentation should it be aded in the future?

Yes, I agree, there probably needs to be some more consideration here. If we don't want to load rdkafka's instrumentation then we should probably load some instrumentation here for producer spans.

As alternatives to this, I can think of the following options:
- Duplicate the spans (there is precedence for this with Faraday and it wrapping of other clients like net/http)
- Have the racecar producer instrumentation used `untraced` so that the rdkafka producer spans are dropped

What do you think?

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (chrisholmes) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb]: done

File: instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec
Comment (chrisholmes) [reply instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec:30]: I'm not sure why it's like this actually. I'll update it depending on the conversation above.

**Comment by ericmustin** [comment]
hey @chrisholmes just ack'ing your followups, plan to catch up here otw. hectic week for me there's been some internal stuff going on at work.

one thing we discussed in the SIG mtg on tuesday was moving this approach to use ASN notifications. I think everyone agreed that if it could be accomplished, it would be preferrable. I'm going to try to see if i can't fix the issues you'd mentioned when you tried the approach, if it is manageable to fix, this would probably be the preferred way forward.

**Comment by chrisholmes** [comment]
> hey @chrisholmes just ack'ing your followups, plan to catch up here otw. hectic week for me there's been some internal stuff going on at work.

thanks @ericmustin!

> one thing we discussed in the SIG mtg on tuesday was moving this approach to use ASN notifications. I think everyone agreed that if it could be accomplished, it would be preferrable. I'm going to try to see if i can't fix the issues you'd mentioned when you tried the approach, if it is manageable to fix, this would probably be the preferred way forward.

Sure, I'm happy for you to try. I should have a working copy somewhere with my experiment too that I can push to a branch though I did it by implementing a customer Racer::Instrumenter rather than hoooking into ASN. For reference, the only real issue I had was with the Racecar's runner not providing [batch message headers to the instrumentation layer](https://github.com/zendesk/racecar/blob/master/lib/racecar/runner.rb#L209). If we're happy to omit links for batch messages then it should be a simply implementation.

**Comment by arielvalentin** [comment]
@chrisholmes Thank you again for your patience and for being willing to try a different approach here. 

How would you feel about submitting an upstream PR to see if @dschierbeck or @bquorning would be willing to include the batch message headers so we are able to use links?

**Comment by chrisholmes** [comment]
@arielvalentin (and @ericmustin), yes, I'm happy to raise a request for this. 

I've been exploring this further and I noticed that using ASN for batch messages runs into the similar issue as described above where the `with_pause` method swallows exceptions before they can get to the instrumentation layer. I expect we will need an additional change to workaround this.

As this might take a while to source out, would you be fine with merging this PR without support for instrumentation of batch message processing? I will push a version with ASN support for single message processing in the meantime.

**Comment by ericmustin** [comment]
@chrisholmes i'd be ok with that. Did not catch up otw, as these things go. will review this evening.

File: instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec
Comment (ericmustin) [inline instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec:28]: nit because it feels like quite the edge case, but i guess i have some misgivings about requiring activesupport explicitly as a dependancy. I'd almost rather just fail the compatibility check if the appropriate active support classes/modules aren't present, ala what active model serializer instrumentation does

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/5d92b37671c59d8aff86d17e8661c3fc25da6a21/instrumentation/active_model_serializers/lib/opentelemetry/instrumentation/active_model_serializers/instrumentation.rb#L23-L25

**Review by ericmustin — APPROVED** [review]
I think this is incredibly close, i'm sorry for all the back and forth here. I had one nit that i'd like to get clarification on (I will ask about this in today's sig mtg and follow up), regarding whats the appropriate way to manage the activesupport dependancy. the prior art in `-contrib` it seems is to not require it as an explicit dependency if possible, and simply include checking for it's existence in a compatibility check. I'm not sure if that's _correct_ though so i'm going to get clarity on that today and follow up. I'm approving this though, with that one caveat.

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (arielvalentin) [inline instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb:38]: @robertlaurin Should this use the OTel Instrumentation Package instead?

```ruby
tracer = :OpenTelemetry::Instrumentation::Racecar.tracer
::OpenTelemetry::Instrumentation::ActiveSupport.subscribe(tracer, process_message.racecar', notification_payload_transform: ->(payload) { #...some logic here 
}) 
```

File: instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec
Comment (chrisholmes) [reply instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec:28]: yes, I was expecting feedback along these lines.

I can't make the SIG, but I will adjust based on the decision there.

Comment (chrisholmes) [reply instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec:28]: 👍 I'll get this pushed later

Comment (chrisholmes) [reply instrumentation/racecar/opentelemetry-instrumentation-racecar.gemspec:28]: right, should be ready now

**Review by ericmustin — APPROVED** [review]

File: instrumentation/racecar/test/opentelemetry/instrumentation/racecar_test.rb
Comment (arielvalentin) [inline instrumentation/racecar/test/opentelemetry/instrumentation/racecar_test.rb:64]: @open-telemetry/ruby-contrib-approvers, we should keep an eye on this to make sure the tests so we can avoid non-deterministic tests. This may be a good use case for a CountDownLatch.

Comment (chrisholmes) [reply instrumentation/racecar/test/opentelemetry/instrumentation/racecar_test.rb:64]: Reading this again, I'd like to tidy this up a little too. I'll introduce CountDownLatch and remove some of the duplication.

**Review by arielvalentin — APPROVED** [review]

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb
Comment (plantfansam) [inline instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/instrumentation.rb:37]: Can we link to what emits this?

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb
Comment (plantfansam) [inline instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb:11]: I believe we're satisfying the `ActiveSupport::Subscriber` [interface](https://api.rubyonrails.org/classes/ActiveSupport/Subscriber.html) here. If so, can we explicitly say so in a comment? IDK if it makes sense to subclass the `Subscriber` class, but that would perhaps be another way of achieving this.

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb
Comment (plantfansam) [inline instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb:42]: I think this instrumentation is adapted from the `ActiveSupport` instrumentation's `SpanSubscriber` ([link](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/22796cb8141b3100c6ce94c25a15e31df3313464/instrumentation/active_support/lib/opentelemetry/instrumentation/active_support/span_subscriber.rb)), so maybe these questions were answered over there. At the risk of being told to RTFM, are we sure that:

1. `start` and `finish` will be called sequentially, with all context.attach calls matched with a context.detach call.
2. `start` and `finish` will be invoked by the same `Thread`.

If user code attaches to a new context between `start` and `finish`, but neglects to detach from that context, then [`detach`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb#L55-L62) will not work as expected. Of course, they'd see an error in their logs, so ... that'll probably help them fix it.

Similarly, if `finish` is called in a different thread than `start`, it'll be using the wrong Context stack, since that's stored on [`Thread.current`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb#L127).

I don't know much about ActiveSupport notifications (or Rdkafka! or Racecar!), so I'm most likely being paranoid here, but thought it was worth asking.

Comment (plantfansam) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb:42]: @ahayworth might have insight, since he worked on a [similar thing](https://github.com/open-telemetry/opentelemetry-ruby/pull/796).

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb
Comment (chrisholmes) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb:12]: Unfortunately, this introduces a ABC rubocop failure

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb
Comment (chrisholmes) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb:42]: TBH, it's not 100% in manual, but `ActiveSupport::Notifications` will call in that order and in the same thread (they leave to subscribers to decide if async is needed). The code isn't terribly complicated [so you could follow on the source](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/notifications.rb)

File: instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb
Comment (plantfansam) [reply instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb:12]: OK. We can always enable/disable the cop but I don't think it's worth the time 😄

File: instrumentation/all/opentelemetry-instrumentation-all.gemspec
Comment (chrisholmes) [reply instrumentation/all/opentelemetry-instrumentation-all.gemspec:53]: I was considering something like that. Has this happened before?

Comment (plantfansam) [reply instrumentation/all/opentelemetry-instrumentation-all.gemspec:53]: Not sure, but maybe @open-telemetry/ruby-contrib-maintainers might know? I don't think there's any problem with it.

Comment (arielvalentin) [reply instrumentation/all/opentelemetry-instrumentation-all.gemspec:53]: Personally speaking I'm ok with adding it. If folks that use all don't want it they should disable it.

I minor bump with the change log/Release notes should be enough of a signal

**Review by plantfansam — APPROVED** [review]
I'd still like to see my open comments resolved at some point, but I don't think it's worth blocking merge. Happy to put this in the wild and see how it fares.

**Comment by chrisholmes** [comment]
@arielvalentin @plantfansam: I've resolved some conflicts and I _think_ that I've switched over to the new CI structure. Please let me know if I've missed anything.

**Review by arielvalentin — APPROVED** [review]

**Pull request merged**
